File: 2019/day4/day4.py
f = open('input', 'r')

def verify(n):
    s = str(n)
    double = False
    for i in range(len(s)-1):
        if int(s[i]) > int(s[i+1]):
            return False
        if int(s[i]) == int(s[i+1]):
            double = True
    return double

def verify2(n):
    s = str(n)
    for c in s:
        if s.count(c) == 2:
            return True
    return False

start = 240920
end = 789857
accept = 0
acc = 0

for i in range(start, end):
    if verify(i):
        accept += 1
        if verify2(i):
            acc += 1

print(verify(111111))
print(verify(223450))
print(verify(123789))
print(accept)

print(verify2(112233))
print(verify2(123444))
print(verify2(111122))
print(acc)

File: 2019/day4/day4.cpp
#include <stdio.h>

bool verify(int n) {
    int i = 100000;
    bool duplicate = false;
    while (i > 1) {
        int cur = (n/i) % 10;
        int nex = (n*10/i) % 10;

        if (cur > nex)
            return false;

        if (cur == nex)
            duplicate = true;

        i /= 10;
    }

    return duplicate;
}

bool verify2(int n) {
    int i = 100000, buckets[10] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
    while (i > 0) {
        int cur = (n/i) % 10;
        buckets[cur]++;
        i /= 10;
    }

    for (int j = 0; j < 10; j++) {
        if (buckets[j] == 2)
            return true;
    }

    return false;
}

int main() {
    int start = 240920;
    int end = 789857;
    int accept = 0, acc = 0;
    for (int i = start; i <= end; i++) {
        if (verify(i)) {
            accept++;
            if (verify2(i)) {
                acc++;
            }
        }
    }

    printf("%d\n", accept);
    printf("%d\n", acc);
}